zepl/contrib/clojure.vim

```diff
@@ -0,0 +1,67 @@
+" Description:  Text formatter for Clojure code.
+" File:         zepl/contrib/clojure.vim
+" Author:       Paul Fernandez <paul4nandez@gmail.com>
+" Help:         :help zepl-clojure
+" Legal:        No rights reserved.  Public domain.
+
+function s:stripLeadingEmptyLines(lines) abort
+   while !empty(a:lines)
+      if a:lines[0] =~# '\m^\s*$'
+         call remove(a:lines, 0)
+      else
+         break
+      endif
+   endwhile
+   return a:lines
+endfunction
+
+function s:stripTrailingEmptyLines(lines) abort
+   let idx = len(a:lines) - 1
+   while idx >= 0
+      if a:lines[idx] =~# '\m^\s*$'
+         call remove(a:lines, idx)
+         let idx -= 1
+      else
+         break
+      endif
+   endwhile
+   return a:lines
+endfunction
+
+function s:normalizeIndents(lines) abort
+   let depth = len(matchstr(a:lines[0], '\m\C^\s*'))
+   return map(a:lines, 'v:val[' . depth . ':]')
+endfunction
+
+function s:processLine(...) abort
+   " Remove trailing newlines.
+   let line = trim(a:2, "\r\n", 2)
+
+   " Remove two leading spaces added by the Clojure repl.
+   return substitute(line, '\m^\s\s', '', '')
```

Comment:
I noticed this behavior in both the Clojure and Leiningen repls. A two-space indent is always added after the first line break.

Before
<img width="116" alt="Screen Shot 2022-05-11 at 4 35 09 PM" src="https://user-images.githubusercontent.com/458926/167968880-1d6bfffa-387f-47e6-8d7d-4ed62cefff6b.png">

After
<img width="109" alt="Screen Shot 2022-05-11 at 4 36 21 PM" src="https://user-images.githubusercontent.com/458926/167968934-4bfb3dc7-0892-4c2f-9ecc-58b60dc3a38b.png">

Reply:
That's odd.  I've never had this problem, so using this formatter actually has the inverse problem.  What OS and Vim version are you using?

| With formatter | Without formatter |
|-----|----|
| <img src="https://user-images.githubusercontent.com/26504626/168667199-c41b9eb1-d8f3-4cd9-a1ac-3aaae406ba07.png" /> | <img src="https://user-images.githubusercontent.com/26504626/168667006-3f85226b-5602-4fc2-91d2-cc37a5eaca02.png" />

I'm fine including this, but maybe there should be a config option for it?

Reply:
I'm using macOS Monterey 12.3.1, and run Vim inside iTerm:

```
$ vim --version
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Apr 20 2022 18:55:37)
macOS version - x86_64
Included patches: 1-4800
Compiled by Homebrew
Huge version without GUI.
```

I'll go back and strip my vimrc to the bare essentials, and also try it with Gvim and on a Ubuntu machine I have lying around. While I'm at it, I'll dig deeper into the mysterious `^M` I mentioned above.

Reply:
I've been pretty busy at work, but today I removed everything except zepl from my vim config and tried it again. I found that the problem is the same in either case, but that my "eliminate two spaces" code only worked because I was always using a two-space indent inside comment forms. I was just covering up the fact that the indentation code that I borrowed from the F# formatter wasn't actually doing anything.

I'll come back to this when I have more time and give it a fresh look.

Reply:
Sure, don't worry, there's no rush.  I'm happy to help where needed, but I am quite busy too.